File: valdi/src/java/com/snap/valdi/utils/PathInterpolator.kt
package com.snap.valdi.utils

import android.graphics.Path
import android.graphics.PathMeasure

class PathInterpolator {

    val empty: Boolean
        get() = pathMeasures.isEmpty()

    private val pathMeasures = arrayListOf<PathMeasure>()
    private var totalLength = 0.0f
    private val outputPath = Path()

    fun setPath(path: Path) {
        reset()

        // Android does not expose the PathContour API, so there is no way to efficiently
        // index all the contours in one iteration. We thus create one PathMeasure to iterate over
        // the contours, we then generate a Path representing the entire contour, on which we
        // create a PathMeasure just for that contour, so that we can later iterate on it.
        val pathMeasure = PathMeasure(path, false)

        do {
            val contourLength = pathMeasure.length

            val contourPath = Path()
            pathMeasure.getSegment(0.0f, contourLength, contourPath, true)

            pathMeasures.add(PathMeasure(contourPath, false))

            this.totalLength += contourLength
        } while (pathMeasure.nextContour())
    }

    fun reset() {
        totalLength = 0.0f
        pathMeasures.clear()
    }

    fun interpolate(start: Float, end: Float): Path {
        outputPath.reset()

        val absoluteStart = start * totalLength
        val absoluteEnd = end * totalLength

        var currentStart = 0.0f
        for (pathMeasure in pathMeasures) {
            val currentLength = pathMeasure.length
            val currentEnd = currentStart + currentLength

            if (currentStart >= absoluteEnd) {
                break
            }

            val relativeStart = Math.min(Math.max(absoluteStart - currentStart, 0.0f), currentLength)
            val relativeEnd = Math.min(absoluteEnd - currentStart, currentLength)

            if (relativeStart != relativeEnd) {
                if (!pathMeasure.getSegment(relativeStart, relativeEnd, outputPath, true)) {
                    break
                }
            }

            currentStart = currentEnd;
        }
        return outputPath
    }

}

File: snap_drawing/src/snap_drawing/cpp/Utils/PathInterpolator.cpp
//
//  PathInterpolator.cpp
//  snap_drawing
//
//  Created by Simon Corsin on 4/20/22.
//

#include "snap_drawing/cpp/Utils/PathInterpolator.hpp"

namespace snap::drawing {

PathInterpolator::PathInterpolator(const Path& path) {
    SkContourMeasureIter iter(path.getSkValue(), false);

    sk_sp<SkContourMeasure> measure;
    while ((measure = iter.next())) {
        _totalLength += measure->length();
        _contours.emplace_back(std::move(measure));
    }
}

PathInterpolator::~PathInterpolator() = default;

const Path& PathInterpolator::interpolate(Scalar start, Scalar end) {
    _interpolatedPath.reset();

    auto absoluteStart = start * _totalLength;
    auto absoluteEnd = end * _totalLength;

    Scalar currentStart = 0;
    for (const auto& contour : _contours) {
        auto currentLength = contour->length();
        auto currentEnd = currentStart + currentLength;

        if (currentStart >= absoluteEnd) {
            break;
        }

        auto relativeStart = std::min(std::max(absoluteStart - currentStart, 0.0f), currentLength);
        auto relativeEnd = std::min(absoluteEnd - currentStart, currentLength);

        if (relativeStart != relativeEnd) {
            if (!contour->getSegment(relativeStart, relativeEnd, &_interpolatedPath.getSkValue(), true)) {
                break;
            }
        }

        currentStart = currentEnd;
    }

    return _interpolatedPath;
}

} // namespace snap::drawing
